design/ceph-mon-pv.md

```diff
@@ -0,0 +1,139 @@
+# Ceph monitor PV storage
+
+Currently all of the storage for Ceph monitors (data, logs, etc..) is provided
+using HostPath volume mounts. Supporting PV storage for Ceph monitors in
+environments with dynamically provisioned network block devices (AWS, GCE,
+etc...) will allow monitors to migrate without requiring the monitor state to be
+rebuilt. This is particularly useful in cases where nodes fails or are taken
+offline for maintenance or upgrades.
+
+There are two major components to this design. First, the cluster CRD schema is
+modified to accept a PVC template for defining the storage to be used by
+monitors. And second, monitor pod management will be handled by StatefulSets
+rather than Deployments.
+
+# Monitor storage
+
+The monitor specification in the CRD is updated to include a persistent volume
+claim template that is used to obtain PVs for the monitor data store.
+
+```go
+type MonSpec struct {
+	Count                int  `json:"count"`
+	PreferredCount       int  `json:"preferredCount"`
+	AllowMultiplePerNode bool `json:"allowMultiplePerNode"`
+	VolumeClaimTemplate  *v1.PersistentVolumeClaim
+}
+```
+
+Unlike the storage requirements of OSDs (xf: StorageClassDeviceSets) it is
+reasonable to use default requirements for monitors (e.g. 5-10 GB). This means
+that a minimal claim may be specified using only the requested storage class in
+which case default resource requests will be used.
+
+```yaml
+volumeClaimTemplates:
+- spec:
+    storageClassName: gp2-ebs
+...
+volumeClaimTemplates:
+- spec:
+    storageClassName: local-storage
```

Comment:
If the storageClassName is blank would we use the default storage class?

Reply:
Presumably we should allow a default storage class--but I suspect we want to force users to make that decision themselves. That is, Rook's default shouldn't be the default storage class. I'm not really sure what hidden weird ux issues there are around this.

Reply:
It should match user expectations if we follow the K8s pattern of an empty storage class name to use the default.